src/integration.ts

```diff
@@ -301,20 +302,143 @@ export function makeIntegration<K extends string, F extends AnyFunction>(
 
 export type CallContext = ASL | VTL | Function<any, any> | EventBus<any>;
 
-/**
- * Dive until we find a integration object.
- */
 export function findDeepIntegrations(
-  expr: FunctionlessNode
-): IntegrationCallExpr[] {
-  const integrations: IntegrationCallExpr[] = [];
-  visitEachChild(expr, function find(node: FunctionlessNode): FunctionlessNode {
-    if (isIntegrationCallExpr(node)) {
-      integrations.push(node);
+  ast: FunctionlessNode
+): CallExpr<ReferenceExpr>[] {
+  const nodes: CallExpr<ReferenceExpr>[] = [];
+  visitEachChild(ast, function visit(node: FunctionlessNode): FunctionlessNode {
+    if (isCallExpr(node)) {
+      const integrations = tryFindIntegrations(node.expr);
+      if (integrations) {
+        nodes.push(
+          ...integrations.map((integration) =>
+            node.fork(
+              new CallExpr(
+                new ReferenceExpr("", () => integration),
+                node.args.map((arg) => arg.clone())
+              )
+            )
+          )
+        );
+      }
     }
-    return visitEachChild(node, find);
+
+    return visitEachChild(node, visit);
   });
-  return integrations;
+
+  return nodes;
+}
+
+/**
+ * A bottom-up algorithm that determines the ONLY {@link Integration} value that the {@link node}
+ * will resolve to at runtime.
+ *
+ * If the {@link node} resolves to 0 or more than 1 {@link Integration} then `undefined` is returned.
+ *
+ * **Note**: This function is an intermediate helper until we migrate the interpreters to be more general
+ * (likely by migrating to top-down algorithms, see https://github.com/functionless/functionless/issues/374#issuecomment-1203313604)
+ *
+ * @param node the node to resolve the {@link Integration} of.
+ * @returns the ONLY {@link Integration} that {@link node} can resolve to, otherwise `undefined`.
+ */
+export function tryFindIntegration(
+  node: FunctionlessNode
+): Integration | undefined {
+  const integrations = tryFindIntegrations(node);
+  if (integrations?.length === 1) {
+    return integrations[0];
+  }
+  return undefined;
+}
+
+/**
+ * A bottom-up algorithm that determines all of the possible {@link Integration}s that a {@link node}
+ * may resolve to at runtime.
+ *
+ * ```ts
+ * declare const table1;
+ * declare const table2;
+ *
+ * const tables = [table1, table2];
+ *
+ * const a = table1;
+ *    // ^ [table1]
+ *
+ * for (const a of tables) {
+ *   const b = a;
+ *          // ^ [table1, table2]
+ *
```

Comment:
Is this implemented? I don't see any code that could return an array of more than 1

Reply:
Not yet, no. I implemented only the cases I needed for the tests to pass - incremental progress. I added the comments as I was thinking it through.

Reply:
We would need to update the interpreter logic to support this case so it didn't matter.